File: 2-DataStructures/1-Arrays/arrays-tests.js
// 4 items in a 32 bits system = 16 bytes of storage
const arr1 = [1, 'b', 'c', 3]

arr1.push('dd'); // O(1) - adding at the end of the array

arr1.pop(); // O(1) - removing at the end of the array
arr1.pop();

arr1.unshift('aa'); // O(n) - adding at the beginning of the array
arr1.unshift('bb');
arr1.shift(); // O(n) - removing at the beginning of the array

arr1.splice(2, 0, 'cc'); // O(n) - inserting at a specific index

console.log('arr1: ', arr1);
console.log('arr1[2]: ', arr1[2]); // O(1) - accessing an element by index

// JS, Python (lists), Java (ArrayLists) - Dynamic Arrays
// C, C++ - Static Arrays
// Dynamic arrays expand as you add more elements
// Static arrays have a fixed size

// implementing an array
class CustomizedArray {
  constructor() {
    this.length = 0;
    this.data = {};
  }

  get(index) {
    return this.data[index]
  }

  push(item) {
    this.data[this.length] = item;
    this.length++;
    return this.length;
  }

  pop() {
    delete this.data[this.length - 1];
    this.length--;
    return this.length;
  }

  delete(index) {
    this.removeItem(index);
    return this.data;
  }

  removeItem(index) {
    for (let i = index; i < this.length - 1; i++) {
      this.data[i] = this.data[i + 1]
    }
    delete this.data[this.length - 1];
    this.length--;
  }

  shiftItems(index) {
    for (let i = index; i < this.length - 1; i++) {
      this.data[i] = this.data[i + 1]
    }
  }
}

const arr2 = new CustomizedArray();
arr2.push('aaa');
arr2.push('bbb');
arr2.push('ccc');
arr2.push('ddd');
arr2.push('eee');
console.log('arr2: ', arr2);
console.log('arr2[1]: ', arr2.get(1));
arr2.pop();
console.log('arr2: ', arr2);
arr2.delete(1);
console.log('arr2: ', arr2);
arr2.shiftItems(1);
console.log('arr2: ', arr2);

File: 2-DataStructures/1-Arrays/arrays-tests.py
# Arrays definition: it's a collection of items stored at contiguous memory locations.
# The idea is to store multiple items of the same type together.
# This makes it easier to calculate the position of each element by simply adding an offset to a base value,
# i.e., the memory location of the first element of the array (generally denoted by the name of the array).
# The base value is index 0 and the difference between the two indexes is the offset.
# Time complexity: O(1) for access, O(n) for search, O(n) for insertion, O(n) for deletion
# Space complexity: O(n)

# Python lists are arrays, but they are not fixed in size.
# if you want to use a real array in Python, you should use the array module. Problem: it only supports a single datatype.
# or you can use the NumPy package, which is a much better alternative. Problem: it's not a part of the standard Python library.

my_list = [1, 5, 10, 4, "hello", 3.14, True, [1, 2, 3]]
print(my_list)
print("Length: ", len(my_list))

my_list[0] = 212
for item in my_list:
    print(item)

# del operator to delete an item at a specific index. Time complexity: O(n)
del my_list[0]
print(my_list)
print("Length: ", len(my_list))

# append() to add an item at the end of the list. Time complexity: O(1)
my_list.append("new item")
print(my_list)

# print the last 2 items
print(my_list[-2:])
print("Length: ", len(my_list))

print(25*"-")

# insert() to add an item at a specific index. Time complexity: O(n)
my_list.insert(2, "new item")
print(my_list)
print("Length: ", len(my_list))

list1 = [1, 2, 3]
list2 = [4, 5, 6]

result = list1 + list2  # combine the two items. Time complexity: O(n)
print(result)

# add the second list to the first one. Time complexity: O(n)
list1.extend(list2)
list1.append("test")
print(list1)
print(list2)

result2 = list1.copy()  # copy the list. Time complexity: O(n). Space complexity: O(n)
print(result2)

# remove the first occurrence of the item. Time complexity: O(n)
list1.remove(2)
list1.pop()  # remove the last item. Time complexity: O(1)
print(list1)

list1.reverse()  # reverse the list. Time complexity: O(n)
print(list1)
list1.sort()  # sort the list. Time complexity: O(n log n)
print(list1)

print(25*"-")

# list comprehension: a concise way to create lists.
# It consists on creating a new list from an existing one, based on a condition.
# Syntax: [expression for item in list if condition]

numbers = [1, -6, 3, 5, -2, 8, 0, -1]

new_list = [x for x in numbers if x > 0]

print("newList: ", new_list)

new_list.clear()
for num in numbers:
    if num % 2 == 0:
        new_list.append(num)

print("newList: ", new_list)

new_list.clear()
new_list = [num for num in numbers if num % 2 == 0]
print("newList: ", new_list)
